### ISSUE TITLE
unexpected cabal-version higher than 2.2 cannot be specified as a range

### ISSUE THREAD

**Issue by dukedave** [sol/hpack#292]
If I try to specify `cxx-options: -std=c++11`  I get:

```

- 19:23:
unexpected cabal-version higher than 2.2 cannot be specified as a range. See https://github.com/haskell/cabal/issues/4899
expecting ".", "-", white space, "&&" or "||"

>= 2.2
```

The referenced issue is https://github.com/haskell/cabal/issues/4899, but I don't know enough to know how that relates to this this. I _guess_ it's something sufficiently new in cabal that it isn't supported in hpack, and to that end:

I see this in the the hpack readme:

> Note: Hpack does not require you to specify a cabal-version manually. When generating a .cabal file, Hpack sets the cabal-version automatically based on the features that are used.

Perhaps it's worth documenting what the latest known working `cabal-version` is (and perhaps warning when going beyond it, as a I presume `2.2` is?).

**Comment by tfausak** [comment]
It used to be that the Cabal file had to have something like this:

```
cabal-version: >= 1.10
```

As of Cabal 2.0, the following syntax was allowed:

```
cabal-version: 2.0 -- essentially means ">= 2.0"
```

As of Cabal 2.2, that syntax was required. hpack needs to choose the appropriate syntax for the given version of Cabal.

**Comment by sol** [comment]
The designated workaround for this would be:

```yaml
verbatim:
  cabal-version: 2.2
```

However, Cabal won't accept that neither as `cabal-version` is now required to be at the beginning of the file.

Not even comments are allowed before `cabal-version` anymore.  This is problematic as `hpack` puts a header comment at the beginning of the file.  It's not yet clear to me what's the best way to address this.

**Comment by sol** [comment]
As I understand it, the following steps will be required to address this:

1. We have to put the `cabal-version` at beginning of the generated cabal file, even before the header.  This will make calculating the file hash more convoluted; it's ugly but solvable.
1. When `cabal-version` is `2.1` or higher, we have to skip the `>=` when rendering.

**Comment by dukedave** [comment]
Also if I manually edit the generated `.cabal` to put `cabal-version: 2.2` at the top, I then get:

```
unexpected Unknown SPDX license identifier: 'BSD3' Do you mean BSD-3-Clause?
```

If I also manually set `license:  BSD-3-Clause` then it works.

I'm not sure other licenses have also been changed.
Shall I open a separate issue for this, or keep it all on this one?

**Comment by sol** [comment]
@dukedave yes, separate issue please.  It may also make sense to make the cabal devs aware that those breaking changes impact users.  While I'm able to capture the motivation behind the `cabal-version` change, the breaking change to `license` is hard to justify by my book (but of course, I may not have the full picture here).

**Comment by sol** [comment]
@tfausak we still have to address this in `hpack`, as Hpack users usually do not set `cabal-version` explicitly.

**Comment by sol** [comment]
Unrecognized license used to be a warning, now it's an error (even though confusingly it says warning in the output).

So the license change are actually two breaking changes, firstly warnings are errors now, secondly e.g. `BSD3` is not accepted anymore.

*no comment*

**Comment by sol** [comment]
Btw, should it be `spdx-license`? `spdx` defines more than just license I assume?

**Comment by quasicomputational** [comment]
Hmm, `spdx-license` is better, yes. I don't know if it's worth trying to be too smart here: if people are on `cabal-version: 2.2` or later, can we tell them to just use `spdx-license` rather than `license`? Detecting that in hpack would be nice, but I don't think it can be perfect in the presence of `verbatim` and fields we don't want to parse too deeply.

**Comment by sol** [comment]
Hmm, from a technical point of view this sounds like a solid solution.

From a users point of view I'm still puzzled whether asking users to change `license` to `spdx-license` and `BSD3` to `BSD-3-clause` as soon as they use e.g. `cxx-options` is justified.  They are semantically the same and odds are that many users just don't care and just want to hack on stuff and continue build their packages.

This whole SPDX change mainly matters for GPL licensed stuff, but most Haskell packages are either in-house or BSD3/MIT anyway.

**Comment by sol** [comment]
Bottom line: I'm still tempted to map at least some of the license to SPDX transparently.

I think it's safe to do so where there is a one-to-one correspondence between traditional cabal and SPDX, which is e.g. the case for `MIT`/`BSD`.

For e.g. `GPL2` it's less clear what to do, as it could mean `GPL-2.0-only` or `GPL-2.0-or-later`.  Apparently Hackage maps to `GPL-2.0-only`, so we could follow that.

**Comment by sol** [comment]
One more thing regarding licenses: I haven't read this in the official spec, but some people say SPDX license are case insensitive.  However, Cabal parses them cases sensitive.

**Comment by tfausak** [comment]
https://spdx.org/spdx-license-list/matching-guidelines:
> 4\. Capitalization
>
> 4.1 Purpose:  To avoid the possibility of a non-match due to lower case or upper case letters in otherwise the same words.
>
> 4.1.1 Guideline:  All upper case and lower case letters should be treated as lower case letters. Templates do not include markup for this guideline.

**Comment by sol** [comment]
Thanks @tfausak. Do we want to shim this in `hpack`?

**Comment by sol** [comment]
I opened https://github.com/sol/hpack/issues/296 for the remaining issue with SPDX license identifiers (case sensitive matching).

**Comment by sol** [comment]
@dukedave released to Hackage as `0.29.0`.  Please test!
